File: greedy/destroying_asteroids.2126.cpp
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    bool asteroidsDestroyed(int mass, vector<int>& asteroids) {
        std::sort(asteroids.begin(), asteroids.end());
        long long total_mass = mass; // use long long to avoid overflow
        for (int m : asteroids) {
            if (total_mass < m) return false;
            total_mass += m;
        }
        return true;
    }
};

File: greedy/destroying_asteroids.2126.py
from typing import List

# O(n) time
# O(1) space
class Solution:
    def asteroidsDestroyed(self, mass: int, asteroids: List[int]) -> bool:
        asteroids.sort()
        for i, m in enumerate(asteroids):
            if mass < m:
                return False
            mass += m
        return True